New plugin system for heat flux boundary conditions

gassmoeller opened the pull request:
This is a proposal for a new plugin system for heat flux boundary conditions for the temperature equation. I need to look over this one more time to look for copy-paste errors, and so far the boundary condition is not actually used. This is just the structure to allow @jdannberg and @jonpkay to continue working on it.

jdannberg commented:
This would be the syntax for the input file:

```
subsection Boundary heat flux model
  set Fixed heat flux boundary indicators = bottom, top
  set Model name = function

  subsection Function
    set Function expression = if(y>0.5,1,-1)
  end
end
```

bangerth commented:
On 06/20/2018 05:16 PM, Juliane Dannberg wrote:
> |set Fixed heat flux boundary indicators = bottom, top|


Shouldn't this be in the same section where we declare the kinds of boundary 
conditions, i.e., in section "Boundary temperature model"?

jdannberg commented:
We also have `Boundary velocity model` and `Boundary traction model`, so I think it makes sense to have the same distinction for the energy equation.

bangerth commented:
We should definitely have the new section. My question is simply where exactly you declare which boundary has what type of boundary condition. Where do you currently say which boundaries are Dirichlet and which are Neumann?

gassmoeller commented:
When we removed the Model settings subsection we moved the specification of the boundary indicators into the separate subsection. There is no central place for the boundary conditions any more, if you prescribe dirichlet boundary conditions for the velocity you do it in `Boundary velocity model`, if you prescribe Neumann boundary conditions you do it in `Boundary traction model`. The same structure should hold for the temperature.

jdannberg commented:
Note to self: we have to add an Assert that makes sure we don't use the same boundary indicator for both temperature and heat flux boundaries.

jdannberg commented:
This should be ready for review now.

tjhei commented on source/boundary_heat_flux/function.cc:
```diff
@@ -0,0 +1,178 @@
+          else
+            {
+              AssertThrow(false, ExcNotImplemented());
+              return std::vector<Tensor<1,dim> > (n_evaluation_points,numbers::signaling_nan<Tensor<1,dim> >());
```
is this line needed to compile? I would guess returning heat_flux below is enough.

tjhei commented on source/boundary_heat_flux/function.cc:
```diff
@@ -0,0 +1,178 @@
+    void
+    Function<dim>::update()
+    {
+      // we get time passed as seconds (always) but may want
```
not "passed"?!

jdannberg replied:
What do you mean, not "passed"? Again, this is just copied from all of the other plugins. Should I just make a separate pull request to change it for all of them? And to what?

tjhei replied:
there is no argument that is "passed". Anyways, we can fix that later.

tjhei commented on source/boundary_heat_flux/function.cc:
```diff
@@ -0,0 +1,178 @@
+                                             "function",
+                                             "Implementation of a model in which the boundary "
+                                             "heat flux is given in terms of an explicit formula "
+                                             "that is elaborated in the parameters in section "
```
elaborated -> given

jdannberg replied:
Same here. Change it everywhere?

tjhei commented on source/simulator/parameters.cc:
```diff
@@ -1204,7 +1234,7 @@ namespace aspect
       AssertThrow(Utilities::has_unique_entries(nullspace_names),
                   ExcMessage("The list of strings for the parameter "
-                             "'Model settings/Remove nullspace' contains entries more than once. "
+                             "'Nullspace removal/Remove nullspace' contains entries more than once. "
```
nice find!

tjhei commented on source/simulator/parameters.cc:
```diff
@@ -1668,6 +1698,27 @@ namespace aspect
+      catch (const std::string &error)
+        {
+          AssertThrow (false, ExcMessage ("While parsing the entry <Boundary heat flux model/Fixed heat flux "
+                                          "boundary indicators>, there was an error. Specifically, "
```
I don't like "there was an error": can you try to rephrase?

jdannberg replied:
That's how it's done for all the other boundary indicators (Free surface, traction, etc.). Do you want me to change all of them? Maybe we just leave it as is?

tjhei replied:
then leave as is.

bangerth commented:
On 06/21/2018 12:13 PM, Rene Gassmoeller wrote:
> When we removed the Model settings subsection we moved the specification of 
> the boundary indicators into the separate subsection. There is no central 
> place for the boundary conditions any more, if you prescribe dirichlet 
> boundary conditions for the velocity you do it in |Boundary velocity model|, 
> if you prescribe Neumann boundary conditions you do it in |Boundary traction 
> model|. The same structure should hold for the temperature.

Hm, I did not pay attention to this at the time. I think it would have been 
nicer if we had had a central place where you say what boundaries have what 
kinds of boundary conditions, and then separate sections that describe the 
details. But that's too late now, so let's not worry about it here.
